// @alexamies/chinesedict-js/index.js
/**
 * Licensed  under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
/**
 * A module that encapsulates a Chinese-English dictionary, including finding
 * words in sections of Chinese text.
 */
export { BasicDictionaryBuilder } from "./src/BasicDictionaryBuilder";
export { DictionaryCollection } from "./src/DictionaryCollection";
export { DictionaryEntry } from "./src/DictionaryEntry";
export { DictionaryLoader } from "./src/DictionaryLoader";
export { DictionarySource } from "./src/DictionarySource";
export { DictionaryView } from "./src/DictionaryView";
export { DictionaryViewConfig } from "./src/DictionaryViewConfig";
export { PlainJSBuilder } from "./src/PlainJSBuilder";
export { Term } from "./src/Term";
export { TextParser } from "./src/TextParser";
export { WordSense } from "./src/WordSense";
